README.md
<div align="center">
  <img src="https://i.imgur.com/rIPQRxz.png" alt="To Do List - Vanilla JavaScript (localStorage)">
  <h1>JavaScript (localStorage)</h1>
</div>

### What is this?

This project is my personal attempt to test and showcase CRUD (Create, Read, Update, Delete) skills using vanilla JavaScript and localStorage. The To-Do List Web App allows users to manage multiple to-do lists with their tasks. 

This web app allows you to create a group to-do list good for different kind of task categories. The task then saved in your browser's localStorage, so it won't be lost when you reload the page. However, clearing your browser's data will also delete the saved to-do list. To make the most of this web app, I recommended to use it with a browser extension that changes the new tab page.

<kbd><img src="https://i.imgur.com/bfEnsLn.png" alt="To Do List UI Preview"></kbd>

### Features

- Create atleast 1000 list and can create multiple task for each list.
- Displays time and date with customizable background image.
- Light and Dark theme with customizable accent color.
- Supports mobile and large screen sizes.

Note: localStorage has a limit of 5MB and clearing your browser's data will also delete the storage. To make the most of this project, it is recommended to use it with a browser extension that changes the new tab page.

### Installation

To install the project, clone the repository to your local machine using Git or download the project as a ZIP file and extract it, then open the `index.html` file in your web browser. There are no additional dependencies or installation steps required.

### Changelog

Go to [Changelog.md](https://github.com/MinecraftJohn/to-do-list/blob/release/CHANGELOG.md) to see previous changes or [Dev-Changelog](https://github.com/MinecraftJohn/to-do-list/blob/development/dev-CHANGELOG.md) for notable development changes.


CHANGELOG.md
# To Do List Changelog
All notable changes to this project will be documented in this file. Dates are displayed in UTC +8.


### v1.4.2
> August 7, 2023
- Fixed deleting the last list do not update elements.

### v1.4.1
> June 9, 2023
- Fixed darkmode switch button not working properly.

### v1.4.0
> April 10, 2023
- Separated the script files to organize and reduce line of codes.
- Page now auto update content, instead of force reload the page.
- Added more screen sizes support for better user experience.
- Fixed bugs on button including on mobile view.
- Removed button being disabled when client opens a menu.
- Added feature where clicking outside of menu closes the menu.
- Added closing animation to all menu.
- Menus now look different on mobile view to support small screen height.
- Fixed logo not resizing on small screen size.
- Fixed last update date not showing the exact result on about menu.
- Fixed autofocus on input element not working sometimes.
- The last list will automatically be selected when deleting a list.
- Added delete message menu/confirmation dialogue when deleting a list or task.
- Selecting a list while on mobile view will now automatically navigate you to task section.
- Updated the logo color to the new default accent color. `assets/svg/favicon.svg`.
- Tab title now show the current selected list.
- Fixed error and issues, optimized codes.

### v1.3.0
> March 17, 2023
- Added the ability to delete a todo group.
- Added the ability to rename a todo group.
- Changed the default theme and accent to `Light Mode` & `#005366`.
- The new default accent color was came from Bing Chat precise mode that also match most of the default wallpaper color that's why I chose it :).
- Replaced green accent `#177d1f` to dark blue green `#005366`.
- Added a closing animation to account modal form.
- Added media query for mobile view support.
- Added a burger menu when client is on mobile view.
- Fixed line divider being shrink on a flex parent.
- Disabled user select on list button at `list_container`.
- Changed `.list_container` height to `32px`.
- Converted `innerHTML` that used string only into `innerText` to avoid bug exploit.
- Disabled button text user selection.
- Added a menu for todo selected in todo header.
- Change main element width to `896px`.

> The new default accent color was came from Bing Chat precise mode that also match most of the default wallpaper color that's why I chose it :).

### v1.2.0
> February 19, 2023
- Added the ability to rename a task instantly.
- Added the ability to mark a task as completed.
- Added the ability to delete a task.
- Remove empty list icon and changed with text only.
- Changed `add_task_btn`style properties and now separated to task list.
- Changed default font family to `Segoe UI Variable Text`.
- Converted SVG buttons to `Segoe Fluent Icon` icons.
- Changed `New list` button text and modal title to `Add a list`.
- Changed `.list_container` height to `32px`.
- Removed empty data message on to do task list.
- Fixed list container element showing a scrollbar even if the container is not overflowed yet on Firefox.
- Changed the list localStorage value from "" to "`[]`".
- Convert innerHTML that is using `+` string concatenation into `${...}` template literals.
- Now able to display the todo list.
- Fixed todo section makes its child that has 100% height overflow issue.

### v1.1.12
> January 29, 2023
- Improved the name validation on the edit account form. It now supports first names only, the dash character ("-"), automatically removes unwanted whitespace, and automatically capitalizes the first letter of each word.
- Disabled autocomplete on the edit account name input.
- Changed to autofocus list and todo input.
- Changed the "Choose image from internet" button in the settings modal to a mini form.
- Modified some grammar in the about project modal.
- Added a feature where the text color on an accent color element changes based on the luminance of the element.
- Changed the data being saved in localStorage from a number (0 or 1) to a boolean (true or false).
- Added a Github API-based feature that displays the last updated date on the about project modal. It will automatically update the date based on the last commit on my To Do Github repository in a `Month dd, yyyy` format.
- Added a fallback color to the background image blocks in the settings modal in case they fail to load.
- Added an active/disable state effect to buttons.
- Added a response status message on custom image url from settings modal and add list modal.
- Improved add list modal.
- List container now displays the list.
- Added a `todo-selected` localStorage to save the last active selected todo list.

### v1.1.11
> November 2, 2022
- Modified the font family CSS file assets/css/fonts.css to use a local font instead of loading it from online.
- Added a fonts folder for the local fonts in `assets/fonts`.
- Made some minor modifications to `assets/css/library.css` and `assets/css/styles.css`.

### v1.1.10
> October 27, 2022
- Added a form modal for creating a new to-do list.

### v1.1.9
> October 24, 2022
- Made some minor modifications to the empty indicator in the to-do list section and task section.

### v1.1.8
> October 21, 2022
- Modified some CSS styles in `assets/css/library.css` and `assets/css/styles.css`.
- Added a hint/icon to indicate that the data is empty and nothing to show.

### v1.1.7
> October 20, 2022
- Fixed the issue where the edit account modal was blurred due to the `mix-blend-mode` CSS property.

### v1.1.6
> October 17, 2022
- Added a dedicated font family CSS file `assets/css/fonts.css` and changed the global font to 'Segoe UI'.
- Modified some CSS styles in library.css and styles.css.
- Fixed the time and date in the header section and settings modal not working.
- Changed the default profile picture to an inline SVG code.

assets/js/script.js
// ############################################
// #           Variables Section              #
// ############################################
var pageHeader = document.querySelector("#header"),
  headerTime = document.querySelector("#time"),
  headerDate = document.querySelector("#date"),
  greetings = document.querySelector("#greetings"),
  accountProfile = document.querySelector("#account_profile"),
  accountPicture = document.querySelector("#account_picture"),
  username = document.querySelector("#username"),
  pageBody = document.getElementsByTagName("body"),
  pageHtml = document.getElementsByTagName("html"),
  editUserBtn = document.querySelectorAll(".setting_edit_user_btn"),
  settings = document.querySelector("#settings"),
  aboutProject = document.querySelector("#about_project"),
  wallpaper = [
    "https://wallpaperaccess.com/full/1779187.jpg",
    "https://wallpaperaccess.com/full/2027653.jpg",
    "https://wallpaperaccess.com/full/1779176.jpg",
    "https://wallpaperaccess.com/full/218253.jpg",
    "https://wallpaperaccess.com/full/148421.jpg",
    "https://wallpaperaccess.com/full/53106.jpg",
  ],
  lsUsername = localStorage.getItem("username"),
  lsProfile = localStorage.getItem("profile"),
  addListBtn = document.querySelector("#add_list_btn"),
  addListBtn2 = document.querySelector("#add_list_btn2"),
  todoTaskSection = document.querySelector("#todo_section"),
  listContainer = document.querySelector("#list_container"),
  listFooterContainer = document.querySelector("#list_footer_container"),
  listName = document.getElementsByClassName("list_name"),
  listsContainer = document.getElementsByClassName("list_container"),
  menuBtn = document.querySelector("#menu_btn");

// ############################################
// #            Startup Section               #
// ############################################
function bgImage(elmnt) {
  if (localStorage.getItem("backgroundImage") == null) {
    elmnt.style.backgroundImage = `url(${wallpaper[0]})`;
  } else {
    elmnt.style.backgroundImage = `url(${localStorage.getItem("backgroundImage")})`;
  }
}

function clock() {
  var time = new Date(),
    hours = time.getHours(),
    minutes = String(time.getMinutes()).padStart(2, "0"),
    day = time.getDay(),
    month = time.getMonth(),
    date = time.getDate(),
    hours12Format = [12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
    dayArray = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"],
    monthArray = [
      "January",
      "February",
      "March",
      "April",
      "May",
      "June",
      "July",
      "August",
      "September",
      "October",
      "November",
      "December",
    ];
  if (JSON.parse(localStorage.getItem("24hour")) === true) {
    headerTime.innerText = hours + ":" + minutes;
  } else {
    headerTime.innerText = hours12Format[hours] + ":" + minutes;
  }
  headerDate.innerText = dayArray[day] + ", " + monthArray[month] + " " + date;
  if (hours >= 6 && hours <= 11) {
    greetings.innerText = "Good morning,";
  } else if (hours >= 12 && hours <= 17) {
    greetings.innerText = "Good afternoon,";
  } else {
    greetings.innerText = "Good evening,";
  }
}

bgImage(pageHeader);
clock();
setInterval(clock, 1000);

function defaultProfilePicture(elmnt, n, more) {
  elmnt.innerHTML = `
    <svg height="${n}" viewBox="0 0 256 256" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path d="M163.786 104.571C163.786 124.434 147.44 140.643 127.143 140.643C106.846 140.643 90.5 124.434 90.5 104.571C90.5 84.7089 106.846 68.5 127.143 68.5C147.44 68.5 163.786 84.7089 163.786 104.571Z" stroke="currentColor" stroke-width="9"/>
        <path fill-rule="evenodd" clip-rule="evenodd" d="M175.137 192.15C174.831 166.14 153.651 145.15 127.569 145.15C101.486 145.15 80.3067 166.14 80.0005 192.15H71C71.3066 161.17 96.5158 136.15 127.569 136.15C158.621 136.15 183.831 161.17 184.137 192.15H175.137Z" fill="currentColor"/>
    </svg>${more}`;
}

function updateUser() {
  if (lsProfile != null) {
    accountProfile.innerHTML = `<img id="account_picture" src="data:image/png;base64,${lsProfile}" alt="User Profile">`;
  } else {
    defaultProfilePicture(accountProfile, 42, "");
  }
  if (lsUsername != null) {
    username.innerText = lsUsername;
  } else {
    username.innerText = "Account Name";
  }
}

updateUser();

function checkThemeMode() {
  if (localStorage.getItem("darkmode") != null) {
    pageHtml[0].setAttribute("darkmode", localStorage.getItem("darkmode"));
  }
  document.documentElement.style.setProperty("--accentColor", localStorage.getItem("accentColor"));
  document.documentElement.style.setProperty("--accentText", localStorage.getItem("accentText"));
}

checkThemeMode();

function closeMenu(elmnt) {
  var modalContainer = document.querySelectorAll(".modal_container"),
    modalBG = document.querySelectorAll(".modal_bg");
  function closeAnimation() {
    setTimeout(() => {
      document.getElementById(elmnt).remove();
      document.removeEventListener("click", removeMenu);
    }, 200);
    modalContainer[0].removeAttribute("id");
  }
  function removeMenu(event) {
    var clickMenu = modalContainer[0].contains(event.target);
    var clickBG = modalBG[0].contains(event.target);
    if (!clickMenu && clickBG) {
      closeAnimation();
    }
  }
  document.addEventListener("click", removeMenu);
  document.querySelector(".close_btn").onclick = closeAnimation;
  setTimeout(() => {
    modalContainer[0].setAttribute("id", "modal_container");
  }, 0);
}

function focusInput(elmnt) {
  elmnt.focus();
  elmnt.setSelectionRange(elmnt.value.length, elmnt.value.length);
}

// ############################################
// #           Mobile View Section            #
// ############################################
menuBtn.onclick = () => {
  if (document.querySelector("#menu_btn_checkbox").checked) {
    menuBtn.innerHTML = `&#xe700;<input type="checkbox" id="menu_btn_checkbox" name="menu_btn_checkbox">`;
    todoTaskSection.classList.replace("hidden", "flex");
    listContainer.classList.add("mobile_menu");
    listFooterContainer.classList.add("mobile_menu");
  } else {
    menuBtn.innerHTML = `&#xe711;<input type="checkbox" id="menu_btn_checkbox" name="menu_btn_checkbox" checked>`;
    todoTaskSection.classList.replace("flex", "hidden");
    listContainer.classList.remove("mobile_menu");
    listFooterContainer.classList.remove("mobile_menu");
  }
};

window.matchMedia("(max-width: 700px)").onchange = () => {
  if (window.matchMedia("(max-width: 700px)").matches) {
    if (document.querySelector("#menu_btn_checkbox").checked) {
      todoTaskSection.classList.replace("flex", "hidden");
    }
  } else {
    if (document.querySelector("#menu_btn_checkbox").checked) {
      todoTaskSection.classList.replace("hidden", "flex");
    }
  }
};

// ############################################
// #            Header TAB Section            #
// ############################################
function updateTAB() {
  const storage = localStorage.getItem("list-selected"),
    title = document.querySelector("title");

  if (storage) {
    title.innerText = `To Do - ${storage.substring(5)}`;
  }
}


assets/js/username.js
editUserBtn[0].onclick = () => {
    var editUserSection = document.createElement("div"),
        fileInputDOM = `<div id="change_picture_btn"><label for="profile">Change</label></div>`,
        profileImgDOM = `<img id="edit_user_account_picture" alt="User Profile" src="data:image/png;base64,${lsProfile}">${fileInputDOM}`;
    pageBody[0].appendChild(editUserSection);
    editUserSection.setAttribute("id", "edit_user_section");
    editUserSection.setAttribute("class", "modal_bg");
    editUserSection.innerHTML = `
        <form class="modal_container" autocomplete="off">
            <header class="modal_header">
                <b>Edit profile</b>
                <i class="close_btn">&#xe8bb;</i>
            </header>
            <div class="line_dividerX"></div>
            <main class="form_body">
                <div id="edit_user_account_profile"></div>
                <input type="file" class="hidden" id="profile" accept=".png, .jpg, .jpeg"/>
                <input type="text" class="input_text" id="change_username" placeholder="Enter a name" maxlength="30">
            </main>
            <div class="line_dividerX"></div>
            <footer class="form_footer">
                <p class="form_error hidden"></p>
                <button id="save_btn">Save</button>
            </footer>
        </form>`;
    var userProfile = document.getElementById("edit_user_account_profile"),
        input = document.getElementById("change_username"),
        saveBtn = document.getElementById("save_btn"),
        errorMsg = document.getElementsByClassName("form_error"),
        profile = document.getElementById("profile"),
        fileReader = new FileReader(),
        modalContainer = document.getElementsByClassName("modal_container");
    function closeAnimation() {
        setTimeout(() => {
            document.getElementById("edit_user_section").remove();
        }, 200);
        modalContainer[0].removeAttribute("id");
    }
    document.getElementsByClassName("close_btn")[0].onclick = closeAnimation;
    setTimeout(() => {
        modalContainer[0].setAttribute("id", "modal_container");
    }, 0);
    if (lsProfile != null) {
        userProfile.innerHTML = profileImgDOM;
    } else {
        defaultProfilePicture(userProfile, 98, fileInputDOM);
    }
    if (lsUsername != null) {
        input.value = lsUsername;
    }
    profile.onchange = () => {
        userProfile.innerHTML = profileImgDOM;
        fileReader.onload = function() {
            document.getElementById("edit_user_account_picture").src = fileReader.result;
        };
        if (profile.files[0]) {
            fileReader.readAsDataURL(profile.files[0]);
        };
    };
    saveBtn.onclick = (e) => {
        const inputValue = input.value.trim().toLowerCase().split(' ').map((s) => s.charAt(0).toUpperCase() + s.substring(1)).join(' ');
        function errorInput(msg) {
            e.preventDefault();
            errorMsg[0].style.display = "block";
            errorMsg[0].innerText = msg;
        }
        if (inputValue.match(/^[a-zA-ZÑñ]+(?: [a-zA-ZÑñ-]+)*$/) && inputValue.length < 33) {
            e.preventDefault();
            localStorage.setItem("username", inputValue);
            if (fileReader.readyState == 2) {
                const base64String = fileReader.result
                    .replace('data:', '')
                    .replace(/^.+,/, '');
                localStorage.setItem("profile", base64String);
                lsProfile = localStorage.getItem("profile");
            }
            lsUsername = localStorage.getItem("username");
            updateUser();
            closeAnimation();
        } else if (inputValue == "") {
            errorInput("Your input name is empty");
        } else {
            errorInput("Number is not allowed");
        }
    };

};

assets/js/create-list.js
function createList() {
  var createListSection = document.createElement("div");
  pageBody[0].appendChild(createListSection);
  createListSection.setAttribute("id", "create_list_section");
  createListSection.setAttribute("class", "modal_bg");
  createListSection.innerHTML = `
        <form class="modal_container form_input" autocomplete="off">
            <header class="modal_header">
                <b>Add a list</b>
                <i class="close_btn">&#xe8bb;</i>
            </header>
            <div class="line_dividerX"></div>
            <main class="form_body add_list_body">
                <div class="input_section">
                    <input type="text" id="name_list" class="input_text" placeholder="Enter a name" >
                    <button id="save_btn" disabled>Save</button>
                </div>
                    <p class="form_error hidden">There is an error creating your list.</p>
            </main>
        </form>`;

  var inputField = document.querySelector("#name_list"),
    saveBtn = document.querySelector("#save_btn");

  closeMenu("create_list_section");
  inputField.focus();

  function saveAddList(e) {
    e.preventDefault();

    if (localStorage.getItem("list-last-id") == null) {
      localStorage.setItem("list-last-id", 0);
    }

    localStorage.setItem("list-last-id", parseInt(localStorage.getItem("list-last-id")) + 1);

    const padStartLastId = localStorage.getItem("list-last-id").toString().padStart(4, "0"),
      todos = "#" + padStartLastId + inputField.value.trim().replace(/^\S/, (c) => c.toUpperCase());

    if (localStorage.getItem(todos) == null) {
      localStorage.setItem(todos, "[]");

      saveBtn.setAttribute("disabled", "");

      localStorage.setItem("list-selected", todos);

      loadTodoList();

      eventListContainer();

      document.querySelector(".close_btn").click();
    } else {
      document.querySelector(".form_error").classList.remove("hidden");
    }
  }

  inputField.oninput = () => {
    if (inputField.value == "" || inputField.value.match(/^\s*$/) || inputField.value.length > 32) {
      saveBtn.setAttribute("disabled", "");
    } else {
      saveBtn.removeAttribute("disabled");
    }
  };

  saveBtn.onclick = saveAddList;
}

addListBtn.onclick = createList;

addListBtn2.onclick = createList;


assets/js/setting.js
settings.onclick = () => {
    var settingsSection = document.createElement("div"),
        color = ["#ea3c78", "#fe8d18", "#ffba25", "#1a73e8", "#005366", "#b040bf"];
    pageBody[0].appendChild(settingsSection);
    settingsSection.setAttribute("id", "settings_section");
    settingsSection.setAttribute("class", "modal_bg");
    settingsSection.innerHTML = `
        <form class="modal_container" autocomplete="off">
            <header class="modal_header">
                <b>Settings</b>
                <i class="close_btn">&#xe8bb;</i>
            </header>
            <div class="line_dividerX"></div>
            <main class="modal_body settings_body">
                <div id="background_image_preview">
                    <h1 id="setting_time" class="time_stats" style="font-size: 32px">00:00</h1>
                    <h3 id="setting_date" class="time_stats">Day, Month 00</h3>
                </div>
                <div class="settings_block">
                    <p>Background</p>
                    <div id="block_container">
                        <div class="block img_block" style="background-image:url(${wallpaper[0]})"></div>
                        <div class="block img_block" style="background-image:url(${wallpaper[1]})"></div>
                        <div class="block img_block" style="background-image:url(${wallpaper[2]})"></div>
                        <div class="block img_block" style="background-image:url(${wallpaper[3]})"></div>
                        <div class="block img_block" style="background-image:url(${wallpaper[4]})"></div>
                        <div class="block img_block" style="background-image:url(${wallpaper[5]})"></div>
                        <div class="block img_block block_active" style="display: none"></div>
                    </div>
                    <div class="img_internet_form">
                        <input type="text" placeholder="Enter a link" id="img_internet_input" />
                        <button id="img_internet_save_btn" disabled>Change</button>
                    </div>
                    <p class="form_error" style="display: none"></p>
                </div>
                <div class="settings_block">
                    <p>Accent color</p>
                    <div id="block_container">
                        <div class="block ac_colors" style="background:${color[0]}"></div>
                        <div class="block ac_colors" style="background:${color[1]}"></div>
                        <div class="block ac_colors" style="background:${color[2]}"></div>
                        <div class="block ac_colors" style="background:${color[3]}"></div>
                        <div class="block ac_colors" style="background:${color[4]}"></div>
                        <div class="block ac_colors" style="background:${color[5]}"></div>
                        <div class="block ac_colors block_active" style="display: none"></div>
                    </div>
                    <div>
                        <label for="accent_picker" class="modal_input_label">Custom color</label>
                        <input type="color" id="accent_picker" value="#005366">
                    </div>
                </div>
                <div class="settings_block">
                    <div class="settings_block settings_block_container">
                        <p>Use 24-hour format</p>
                        <label class="switch_outline" for="24hour">
                            <input class="switch_input" type="checkbox" id="24hour"/>
                            <div class="switch_base"></div>
                            <div class="switch_thumb"></div>
                        </label>
                    </div>
                    <div class="settings_block settings_block_container">
                        <p>Dark mode</p>
                        <label class="switch_outline" for="darkmode">
                            <input class="switch_input" type="checkbox" id="darkmode"/>
                            <div class="switch_base"></div>
                            <div class="switch_thumb"></div>
                        </label>
                    </div>
                </div>
            </main>
        </form>
    `;

    closeMenu("settings_section");
    
    // ######################################
    // #      Variables and Functions       #
    // ######################################
    var bgImagePreview = document.querySelector("#background_image_preview"),
        previewTime = document.querySelector("#setting_time"),
        previewDate = document.querySelector("#setting_date"),
        lsBGImage = localStorage.getItem("backgroundImage"),
        imgBlock = document.querySelectorAll(".img_block"),
        darkmodeInput = document.querySelector("#darkmode"),
        accentPicker = document.querySelector("#accent_picker"),
        acColors = document.querySelectorAll(".ac_colors"),
        hourFormatInput = document.getElementById("24hour"),
        lsAcColor = localStorage.getItem("accentColor"),
        enterLink = document.querySelector("#img_internet_input"),
        enterLinkBtn = document.querySelector("#img_internet_save_btn"),
        errorMsg = document.querySelectorAll(".form_error");

    function updateWallpaper(n) {
        changeBGImage(n, wallpaper[n]);
        bgImage(bgImagePreview);
        bgImage(pageHeader);
    }

    function clock() {
        var time = new Date(),
            hours = time.getHours(),
            minutes = String(time.getMinutes()).padStart(2, "0"),
            day = time.getDay(),
            month = time.getMonth(),
            date = time.getDate(),
            hours12Format = [12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
            dayArray = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"],
            monthArray = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
        if (JSON.parse(localStorage.getItem("24hour")) === true) {
            previewTime.innerText = hours + ":" + minutes;
        } else {
            previewTime.innerText = hours12Format[hours] + ":" + minutes;
        }
        previewDate.innerText = dayArray[day] + ", " + monthArray[month] + " " + date;
    }

    function changeBGImage(a, b) {
        blockActive(imgBlock);
        imgBlock[a].classList.add("block_active");
        localStorage.setItem("backgroundImage", b);
    }

    function attPicked(i) {accentPicker.setAttribute("value", i)}

    function updateColor(n) {
        changeAccentColor(color[n], n);
        setAccentText();
    }

    function blockActive(a) {
        a[0].classList.remove("block_active");
        a[1].classList.remove("block_active");
        a[2].classList.remove("block_active");
        a[3].classList.remove("block_active");
        a[4].classList.remove("block_active");
        a[5].classList.remove("block_active");
        a[6].style.display = "none";
    }

    function changeAccentColor(a, b) {
        attPicked(a);
        blockActive(acColors);
        acColors[b].classList.add("block_active");
        document.documentElement.style.setProperty("--accentColor", accentPicker.getAttribute("value"));
        localStorage.setItem("accentColor", accentPicker.getAttribute("value"));
    }

    function checkAccentColor(color) {
        const r = parseInt(color.slice(1, 3), 16);
        const g = parseInt(color.slice(3, 5), 16);
        const b = parseInt(color.slice(5, 7), 16);
        const luminance = (0.2126 * r) + (0.7152 * g) + (0.0722 * b);
        return luminance >= 128;
    }

    function setAccentText() {
        if (checkAccentColor(localStorage.getItem("accentColor"))) {
            document.documentElement.style.setProperty("--accentText", "#1A1A1B");
            localStorage.setItem("accentText", "#1A1A1B");
        } else {
            document.documentElement.style.setProperty("--accentText", "#ffffff");
            localStorage.setItem("accentText", "#ffffff");
        }
    }
    
    // ##########################
    // #      Loaded Data       #
    // ##########################
    bgImage(bgImagePreview);

    if (lsBGImage == null || lsBGImage == wallpaper[0]) {
        imgBlock[0].classList.add("block_active");
    } else if (lsBGImage == wallpaper[1]) {
        imgBlock[1].classList.add("block_active");
    } else if (lsBGImage == wallpaper[2]) {
        imgBlock[2].classList.add("block_active");
    } else if (lsBGImage == wallpaper[3]) {
        imgBlock[3].classList.add("block_active");
    } else if (lsBGImage == wallpaper[4]) {
        imgBlock[4].classList.add("block_active");
    } else if (lsBGImage == wallpaper[5]) {
        imgBlock[5].classList.add("block_active");
    } else {
        imgBlock[6].setAttribute("style", `background-image:url(${lsBGImage})`);
    }

    clock();

    setInterval(clock, 1000);

    if (JSON.parse(localStorage.getItem("24hour")) === true) {hourFormatInput.setAttribute("checked", "")}

    if (JSON.parse(localStorage.getItem("darkmode")) === true) {darkmodeInput.setAttribute("checked", "")}

    if (lsAcColor == color[0]) {
        attPicked(color[0]);
        acColors[0].classList.add("block_active");
    } else if (lsAcColor == color[1]) {
        attPicked(color[1]);
        acColors[1].classList.add("block_active");
    } else if (lsAcColor == color[2]) {
        attPicked(color[2]);
        acColors[2].classList.add("block_active");
    } else if (lsAcColor == color[3]) {
        attPicked(color[3]);
        acColors[3].classList.add("block_active");
    } else if (lsAcColor == color[4] || lsAcColor == null) {
        attPicked(color[4]);
        acColors[4].classList.add("block_active");
    } else if (lsAcColor == color[5]) {
        attPicked(color[5]);
        acColors[5].classList.add("block_active");
    } else {
        attPicked(lsAcColor);
        acColors[6].style.display = "block";
        acColors[6].style.background = lsAcColor;
    }

    // ##########################
    // #   Settings Auto Save   #
    // ##########################
    imgBlock[0].onclick = () => {updateWallpaper(0)};

    imgBlock[1].onclick = () => {updateWallpaper(1)};

    imgBlock[2].onclick = () => {updateWallpaper(2)};

    imgBlock[3].onclick = () => {updateWallpaper(3)};

    imgBlock[4].onclick = () => {updateWallpaper(4)};

    imgBlock[5].onclick = () => {updateWallpaper(5)};

    enterLink.oninput = () => {
        if (enterLink.value == "" || enterLink.value.match(/\s/)) {
            enterLinkBtn.setAttribute("disabled", "");
        } else {
            enterLinkBtn.removeAttribute("disabled");
        }
    };

    enterLinkBtn.onclick = (e) => {
        e.preventDefault();
        fetch(enterLink.value)
        .then(response => {
            if (response.status === 200) {
                changeBGImage(6, enterLink.value);
                bgImage(bgImagePreview);
                imgBlock[6].setAttribute("style", `background-image:url(${enterLink.value})`);
            } else {
                errorMsg[0].removeAttribute("style");
                errorMsg[0].innerText = "Your request was unsuccessful, try again later.";
            }
        }).catch(() => {
            errorMsg[0].removeAttribute("style");
            errorMsg[0].innerText = "Please enter a valid image URL.";
        });
    };

    accentPicker.oninput = (e) => {
        blockActive(acColors);
        attPicked(e.target.value);
        acColors[6].style.display = "block";
        acColors[6].style.background = e.target.value;
        document.documentElement.style.setProperty("--accentColor", accentPicker.getAttribute("value"));
        localStorage.setItem("accentColor", accentPicker.getAttribute("value"));
        setAccentText();
    }

    acColors[0].onclick = () => {updateColor(0)};

    acColors[1].onclick = () => {updateColor(1)};

    acColors[2].onclick = () => {updateColor(2)};

    acColors[3].onclick = () => {updateColor(3)};

    acColors[4].onclick = () => {updateColor(4)};

    acColors[5].onclick = () => {updateColor(5)};

    hourFormatInput.onclick = () => {
        if (hourFormatInput.checked) {
            localStorage.setItem("24hour", true);
        } else {
            localStorage.setItem("24hour", false);
        }
    };

    darkmodeInput.onclick = () => {
        if (darkmodeInput.checked) {
            localStorage.setItem("darkmode", true);
        } else {
            localStorage.setItem("darkmode", false);
        }
        checkThemeMode();
    };
};